devtools-integration-tests/testerina-objectmocking-test/tests/main_test.bal

```diff
@@ -0,0 +1,290 @@
+    if (mockSmtpClient is error) {
+        test:assertEquals(mockSmtpClient.message(), "mock object type 'MockSmtpClientEmpty' should have at least one member function or field declared.");
+    } else {
+        test:assertFail(msg = "Empty user defined object to mock not handled!");
+    }
+}
+
+
+// 1.2) when user-defined object is passed to test:prepare function
+@test:Config { }
+function testUserDefinedMockRegisterCases() {
+    email:SmtpClient mockSmtpClient = <email:SmtpClient>test:mock(email:SmtpClient, new MockSmtpClient());
+    error? result = trap test:prepare(mockSmtpClient).when("send").doNothing();
+    if (result is error) {
+        test:assertEquals(result.message(), "cases cannot be registered to user-defined object type 'SmtpClient'");
+    } else {
+        test:assertFail(msg = "Invalid object for mock object not handled!");
+    }
+}
+
+// 1.3) when the functions in mock is not available in the original
+@test:Config { }
+function testUserDefinedMockInvalidFunction() {
+    email:SmtpClient|error mockSmtpClient = trap <email:SmtpClient>test:mock(email:SmtpClient, new MockSmtpClientFuncErr());
+    if (mockSmtpClient is error) {
+        test:assertEquals(mockSmtpClient.message(), "invalid function 'sendMail' provided");
+    } else {
+        test:assertFail(msg = "Missing function for mock object not handled!");
+    }
+}
+
+// 1.4.1) when the function return types do not match
+@test:Config { }
+function testUserDefinedMockFunctionSignatureMismatch() {
+    email:SmtpClient|error mockSmtpClient = trap <email:SmtpClient>test:mock(email:SmtpClient, new MockSmtpClientSigErr());
+    if (mockSmtpClient is error) {
+        test:assertEquals(mockSmtpClient.message(), "incompatible return type provided for function send()");
+    } else {
+        test:assertFail(msg = "Function signature mismatch for object not handled!");
+    }
+}
+
+// 1.4.2) when the function parameters do not match
+@test:Config { }
+function testUserDefinedMockFunctionSignatureMismatch2() {
+    email:SmtpClient|error mockSmtpClient = trap <email:SmtpClient>test:mock(email:SmtpClient, new MockSmtpClientSigErr2());
+    if (mockSmtpClient is error) {
+        test:assertEquals(mockSmtpClient.message(), "incompatible parameter type provided at position 1 in function send()");
+    } else {
+        test:assertFail(msg = "Function signature mismatch for object not handled!");
+    }
+}
+
+// 1.4.3
+@test:Config { }
+function testUserDefinedMockFunctionSignatureMismatch3() {
+    http:Client|error mockHttpClient = trap <http:Client>test:mock(http:Client, new MockHttpClientSigErr());
+    if (mockHttpClient is error) {
+        test:assertEquals(mockHttpClient.message(), "incorrect number of parameters provided for function get()");
+    } else {
+        test:assertFail(msg = "Function signature mismatch for object not handled!");
+    }
+}
+
+# 2 - Validations for framework provided default mock object
```

Comment:
IMO we don't need to move the validation test cases here, these can be tested at ballerina-lang repo.

Reply:
I have removed the validation portion and I will work on simplifying the test cases